hw1 v1

Ornstein385 opened the pull request.

strazhnick committed:
hw1 v1

strazhnick committed:
hw1 удалил учебный SampleTest

strazhnick committed:
hw1. удалил учебный SampleTest. восстановил случайно удаленный BotTest

strazhnick committed:
hw1. переместил тестовую директорию в нужное место

YuriiKondratov commented on bot/src/main/java/edu/java/bot/configuration/ApplicationConfig.java:
```diff
@@ -1,13 +1,18 @@
+@Getter
+public class ApplicationConfig {
+
+    //@NotEmpty
```
Забыл удалить или зачем-то нужно?

YuriiKondratov commented on bot/src/main/java/edu/java/bot/configuration/ApplicationConfig.java:
```diff
@@ -1,13 +1,18 @@
-import org.springframework.validation.annotation.Validated;
-
-@Validated
-@ConfigurationProperties(prefix = "app", ignoreUnknownFields = false)
```
а чем тебе префикс app не понравился? Это же удобнее, чем писать везде `@Value("${app....}")`.

YuriiKondratov commented on bot/src/main/java/edu/java/bot/controllers/Bot.java:
```diff
@@ -0,0 +1,163 @@
+    public Bot(ApplicationConfig config, LinkDaoInterface linkDao) {
+        this.config = config;
+        this.linkDao = linkDao;
+        this.bot = new TelegramBot(config.getTelegramToken());
```
Тут как-то странно: ты инжектишь linkDao и config, а bot создаешь через new. Давай bot тоже заинжектим.

YuriiKondratov commented on bot/src/main/resources/application.yml:
```diff
@@ -1,5 +1,6 @@
 app:
-  telegram-token: unset
+  telegramToken: 6379106394:AAGx-SZur6HE-KEhKXSyT9MGSJoYnvfYGME
```
удали токен из репозитория, пока никто от имени твоего бота не пошёл деньги вымогать

YuriiKondratov commented on bot/src/main/java/edu/java/bot/controllers/Bot.java:
```diff
@@ -0,0 +1,163 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class Bot {
```
Слишком много всего в одном классе, из-за этого не получится провести нормально юнит-тестирование: не получится отдельно протестировать какой-то небольшой кусочек логики (например, проверку домена). Также возникнут проблемы при внесении изменений: если добавятся новые комманды, то придется слишком много всего редактировать в этом классе, в результате велика вероятность, что что-то сломается.

В общем давай разделим этот класс, выделим необходимые интерфейсы и их реализации, чтобы всё у нас было SOLIDно.

YuriiKondratov commented on bot/src/test/java/edu/java/bot/BotTest.java:
```diff
@@ -0,0 +1,249 @@
+
+        bot.handleUpdate(update);
+
+        verify(linkDao, never()).add(any(Link.class));
```
Здесь ты хочешь протестировать логику проверки домена. Но по факту ты тестируешь очено большое количество другой логики, поэтому если этот тест упадёт, тебе будет понятно, что что-то сломалось, но не будет понятно, что именно сломалось. Если ты разделишь класс Bot на составляющие, то каждую составляющую сможешь протестировать отдельно.

strazhnick committed:
hw1
исправлен класс конфигурации.
теперь инжектим бин для бота.

strazhnick committed:
hw1
переделана структура бота.
отдельные классы для команд.

strazhnick committed:
hw1
переделаны тесты.
удален токен. (предыдущий не был рабочим. он был оставлен для примера)

YuriiKondratov commented on bot/src/main/java/edu/java/bot/telegram/Bot.java:
```diff
@@ -0,0 +1,84 @@
+    private final LinkDaoInterface linkDao;
+    private final CommandKeeper commandKeeper;
+
+    private static final String START = "/start";
```
мне не нравится, что список команд захардкожен - очень тяжело будет новые команды добавлять

YuriiKondratov commented on bot/src/main/java/edu/java/bot/command/Command.java:
```diff
@@ -0,0 +1,21 @@
+
+    public abstract String getDescription();
+
+    protected final TelegramBot bot;
```
По-хорошему команды не должна вообще знать про бота. Единственная её задача - получить update и превратить его в ответное сообщение. Все что дальше с ответом происходит команду волновать не должно. Поэтому давай уберем из команды это поле и логику отправки ответа.

YuriiKondratov commented on bot/src/main/java/edu/java/bot/command/HelpCommand.java:
```diff
@@ -0,0 +1,43 @@
+    @Autowired
+    @Lazy
+    public void setCommandKeeper(CommandKeeper commandKeeper) {
+        this.commandKeeper = commandKeeper;
```
команде нужен комманд кипер и комманд киперу нужна команда.. такая себе идея. Лучше сюда заинжектить список команд а не команд кипер

YuriiKondratov commented:
Как бы я сделал: я бы добавил commandHandler, у которого есть метод `SendMessage handle(Update update)`. Чтобы бот при получении `Update` просто вызывал `execute(handle(update))`. В commandHandler хэндлер я бы заинжектил все команды и спрятал бы в нём логику выбора команды.

strazhnick committed:
переделан принцип вызова команд
